st2common/st2common/services/packs.py

```diff
@@ -102,7 +114,11 @@ def _fetch_and_compile_index(index_urls, logger=None, proxy_config=None):
         index_json = None
 
         try:
-            request = requests.get(index_url, proxies=proxies_dict, verify=verify)
+            request = requests.get(
+                index_url,
+                proxies=proxies_dict if not bypass_proxy else None,
+                verify=verify if not bypass_proxy else True,
```

Comment:
```suggestion
                verify=verify,
```

I still don't understand why adjusting the `verify` logic is needed in this situation.
I thought we were trying to fix `no_proxy`.

Could you perhaps provide more context or an example?

Reply:
https://github.com/StackStorm/st2/blob/224e0e29310a8ebdfba522fb28e59585b8cfc940/st2common/st2common/services/packs.py#L88-L92

Because verify is overwritten with the path to your proxy's ca bundle if `proxy_ca_bundle_path` environment variable is set. Since its outside the loop there could be a chance where you are specifying a ca bundle path but needing to bypass the proxy.

An example might be (without `ca_bundle_path or True`):
```python
proxy_config = {
    "https_proxy": "http://proxy.company.local",
    "http_proxy":  "http://proxy.company.local",
    "no_proxy": "company.local,localhost,my.other.domain",
    "proxy_ca_bundle_path": "/path/to/proxy/ca_bundle"
}

proxies_dict = {
    "https": "http://proxy.company.local",
    "http":  "http://proxy.company.local",
    "no": "company.local,localhost,my.other.domain",
}

# verify is now overwritten from True to '/path/to/proxy/ca_bundle'
verify="/path/to/proxy/ca_bundle"
```

Now we enter loop:
```python
# Iteration 1
verify="/path/to/proxy/ca_bundle" # permanently set

index_url="https://index.stackstorm.com"
bypass_proxy = False

requests.get(
    "https://index.stackstorm.com",
    proxies={
        "https": "http://proxy.company.local",
        "http":  "http://proxy.company.local",
        "no": "company.local,localhost,my.other.domain",
    },
    verify="/path/to/proxy/ca_bundle",
)
```

```python
# Iteration 2
verify="/path/to/proxy/ca_bundle" # permanently set

index_url="https://index.mycompany.local"
bypass_proxy = True

requests.get(
    "https://index.mycompany.local",
    proxies=None
    verify="/path/to/proxy/ca_bundle",
)
```

So in Iteration 2 we would might end up failing because we're still trying to use our proxy's ca bundle, but not using the proxy.

Reply:
I see now, thanks for more context :+1: